examples/DataApi-Voting-poll/template/modules/shared_data/src/main/scala/com/my/currency/shared_data/Main.scala

```diff
@@ -0,0 +1,351 @@
+    val message = "Pool is closed"
+  }
+}
+
+object TypeValidators {
+  def validateIfNewPollExists(maybeState: Option[PollState]): DataApplicationValidationErrorOr[Unit] = {
+    maybeState match {
+      case Some(_) => PollAlreadyExists.asInstanceOf[DataApplicationValidationError].invalidNec
+      case None => ().validNec
+    }
+  }
+
+  def validateIfVotePollExists(maybeState: Option[PollState]): DataApplicationValidationErrorOr[Unit] = {
+    maybeState match {
+      case Some(_) => ().validNec
+      case None => PollDoesNotExists.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+
+  def validateIfOptionExists(maybeState: Option[PollState], option: String): DataApplicationValidationErrorOr[Unit] = {
+    maybeState match {
+      case Some(value) =>
+        val pollValidOptions = value.pollOptions
+        pollValidOptions.get(option) match {
+          case Some(_) => ().validNec
+          case None => InvalidOption.asInstanceOf[DataApplicationValidationError].invalidNec
+        }
+      case None => InvalidOption.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+
+  def validateProvidedAddress(proofAddresses: List[Address], address: Address): DataApplicationValidationErrorOr[Unit] = {
+    if (proofAddresses.contains(address)) {
+      ().validNec
+    } else {
+      InvalidAddress.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+
+  def validateIfUserAlreadyVoted(maybeState: Option[PollState], address: Address): DataApplicationValidationErrorOr[Unit] = {
+    maybeState match {
+      case Some(value) =>
+        val pollVotes = value.votes
+        pollVotes.get(address) match {
+          case Some(_) => RepeatedVote.asInstanceOf[DataApplicationValidationError].invalidNec
+          case None => ().validNec
+        }
+      case None => ().validNec
+    }
+  }
+
+  def validateSnapshotNewPoll(snapshotOrdinal: SnapshotOrdinal, update: NewPoll): DataApplicationValidationErrorOr[Unit] = {
+    if (update.endSnapshotOrdinal < snapshotOrdinal.value.value) {
+      InvalidEndSnapshot.asInstanceOf[DataApplicationValidationError].invalidNec
+    } else {
+      ().validNec
+    }
+  }
+
+  def validatePollSnapshotInterval(lastSnapshotOrdinal: SnapshotOrdinal, currentState: State, pollVote: PollVote): DataApplicationValidationErrorOr[Unit] = {
+    val poll = currentState.polls.get(pollVote.id)
+    poll match {
+      case Some(value) =>
+        if (value.endSnapshotOrdinal < lastSnapshotOrdinal.value.value) {
+          ClosedPool.asInstanceOf[DataApplicationValidationError].invalidNec
+        } else {
+          ().validNec
+        }
+      case None => PollDoesNotExists.asInstanceOf[DataApplicationValidationError].invalidNec
+    }
+  }
+}
+
+object Routes {
+  def routeValidateUpdate(update: PollUpdate, state: State, lastSnapshotOrdinal: SnapshotOrdinal): IO[DataApplicationValidationErrorOr[Unit]] = {
+    update match {
+      case newPoll: NewPoll =>
```

Comment:
This sort of defeats the purpose of "routing" the validations since we still have all the code for each validated type in a single method. The idea would be to organize this code a bit more by splitting the validation for each type into its own function.

Reply:
Updated